src/engraving/rendering/dev/accidentalslayout.cpp

```diff
@@ -0,0 +1,1118 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+#include <cfloat>
+
+#include "accidentalslayout.h"
+#include "layoutcontext.h"
+#include "../paddingtable.h"
+#include "shape.h"
+#include "tlayout.h"
+
+#include "dom/accidental.h"
+#include "dom/chord.h"
+#include "dom/ledgerline.h"
+#include "dom/note.h"
+#include "dom/score.h"
+
+using namespace mu::engraving;
+using namespace mu::engraving::rendering::dev;
+
+AccidentalsLayout::AccidentalsLayoutContext::AccidentalsLayoutContext(const std::vector<Accidental*>& acc,
+                                                                      const std::vector<Chord*>& ch)
+    : allAccidentals(acc), chords(ch)
+{
+    sortTopDown(allAccidentals);
+    stackedAccidentals.reserve(allAccidentals.size());
+    stackedAccidentalsShape.elements().reserve(4 * allAccidentals.size());
+
+    initConstants();
+}
+
+void AccidentalsLayout::AccidentalsLayoutContext::initConstants()
+{
+    m_spatium = allAccidentals.front()->spatium();
+
+    m_verticalAccidentalToAccidentalClearance = 0.15 * m_spatium;
+    m_verticalSharpToSharpClearance = 0.05 * m_spatium;
+    m_verticalAccidentalToChordClearance = 0.10 * m_spatium;
+    m_additionalPaddingForVerticals = 0.1 * m_spatium;
+
+    const MStyle& style = allAccidentals.front()->style();
+    m_accidentalAccidentalDistance = style.styleMM(Sid::accidentalDistance);
+    m_orderFollowNoteDisplacement = style.styleB(Sid::accidentalOrderFollowsNoteDisplacement);
+    m_alignOctavesAcrossSubChords = style.styleB(Sid::alignAccidentalOctavesAcrossSubChords);
+    m_keepSecondsTogether = style.styleB(Sid::keepAccidentalSecondsTogether);
+    m_alignOffsetOctaves = style.styleB(Sid::alignOffsetOctaveAccidentals);
+}
+
+void AccidentalsLayout::layoutAccidentals(const std::vector<Chord*>& chords, LayoutContext& ctx)
+{
+    std::vector<Accidental*> allAccidentals = collectAccidentals(chords);
+
+    if (allAccidentals.empty()) {
+        return;
+    }
+
+    for (Accidental* acc : allAccidentals) {
+        acc->computeMag();
+        TLayout::layoutAccidental(acc, acc->mutldata(), ctx.conf());
+    }
+
+    AccidentalsLayoutContext accidentalsLayoutContext(allAccidentals, chords);
+    doAccidentalPlacement(accidentalsLayoutContext);
+}
+
+std::vector<Accidental*> AccidentalsLayout::collectAccidentals(const std::vector<Chord*> chords)
+{
+    std::vector<Accidental*> accidentals;
+
+    for (Chord* chord : chords) {
+        for (Note* note : chord->notes()) {
+            Accidental* acc = note->accidental();
+            if (!acc) {
+                continue;
+            }
+            acc->setPos(0.0, 0.0);
```

Comment:
This is not ideal... This function not only collects accidentals but also changes some properties. Ideally, it should have only one responsibility (find and return accidentals)